### PR TITLE
add solution

### PR DESCRIPTION
**Pull request by AndrewHudzenko**

### CODE REVIEW

File: src/main/java/mate/academy/dao/UserDao.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import java.util.List;
+import java.util.Optional;
+import mate.academy.model.User;
+
+public interface UserDao {
+    User add(User user);
+
+    Optional<User> get(Long id);
+
+    List<User> getAll();
```
Comment (valentyn-chepiha) [inline src/main/java/mate/academy/dao/UserDao.java:12]:
```suggestion
```
it's not nessesary

File: src/main/java/mate/academy/exception/RegistrationException.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.exception;
+
+public class RegistrationException extends Exception {
+    public RegistrationException(String message) {
+        super(message);
+    }
+
+    public RegistrationException(String message, Throwable cause) {
+        super(message, cause);
+    }
```
Comment (valentyn-chepiha) [inline src/main/java/mate/academy/exception/RegistrationException.java:10]:
```suggestion
```
this is redundant

File: src/main/java/mate/academy/exception/AuthenticationException.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.exception;
+
+public class AuthenticationException extends Exception {
+    public AuthenticationException(String message) {
+        super(message);
+    }
+
+    public AuthenticationException(String message, Throwable cause) {
+        super(message, cause);
+    }
```
Comment (valentyn-chepiha) [inline src/main/java/mate/academy/exception/AuthenticationException.java:10]:
```suggestion
```
this is redundant

File: src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    private UserService userService;
```
Comment (valentyn-chepiha) [inline src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java:14]:
```suggestion
    @Inject
    private UserService userService;
```

File: src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        if (userFromDbByEmail.isEmpty()) {
+            throw new AuthenticationException("Couldn't login");
+        }
+        User user = userFromDbByEmail.get();
+        String hashedPassword = HashUtil.hashPassword(password, user.getSalt());
+        if (user.getPassword().equals(hashedPassword)) {
+            return user;
+        }
+        throw new AuthenticationException("Couldn't login");
+    }
```
Comment (valentyn-chepiha) [inline src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java:28]: Common mistakes
`Create only one condition for throwing AuthenticationException in login() method. You may combine two checks: whether the user has been found by login and does passwords match.`

File: src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Override
+    public User register(String email, String password) throws RegistrationException {
+        if (userService.findByEmail(email).isEmpty()) {
+            User user = new User(email, password);
+            return userService.add(user);
+        }
+        throw new RegistrationException("Couldn't register, account with this email "
+                + email + " is already exist");
+    }
```
Comment (valentyn-chepiha) [inline src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java:38]: what happens if the login or password is null

File: src/main/java/mate/academy/util/HashUtil.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+public class HashUtil {
+    private static final String CRYPTO_ALGORITHM = "SHA-512";
+
+    private HashUtil() {
+
```
Comment (valentyn-chepiha) [inline src/main/java/mate/academy/util/HashUtil.java:11]:
```suggestion
```

**Review by valentyn-chepiha — CHANGES_REQUESTED** [review]

**Review by valentyn-chepiha — APPROVED** [review]

File: src/main/java/mate/academy/exception/AuthenticationException.java (hunk 3 above)
Comment (Sofasmile) [inline src/main/java/mate/academy/exception/AuthenticationException.java:10]: do you need this constructor?

File: src/main/java/mate/academy/exception/RegistrationException.java (hunk 2 above)
Comment (Sofasmile) [inline src/main/java/mate/academy/exception/RegistrationException.java:10]: do you use it?

File: src/main/java/mate/academy/model/User.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        this.salt = salt;
+    }
+
+    @Override
+    public String toString() {
+        return "User{"
+                + "id=" + id
+                + ", email='" + email + '\''
+                + ", password='" + password + '\''
+                + ", salt='" + Arrays.toString(salt) + '\'' + '}';
```
Comment (Sofasmile) [inline src/main/java/mate/academy/model/User.java:67]: we shouldn't add password and salt to this method

File: src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDbByEmail = userService.findByEmail(email);
```
Comment (Sofasmile) [inline src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java:20]:
```suggestion
        Optional<User> userFromDb = userService.findByEmail(email);
```

File: src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        User user = userFromDbByEmail.get();
```
Comment (Sofasmile) [inline src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java:21]: we can get NoSuchElementException here

File: src/main/java/mate/academy/util/HashUtil.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(CRYPTO_ALGORITHM);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalStateException("Couldn't create hash using SHA-512 algorithm", e);
```
Comment (Sofasmile) [inline src/main/java/mate/academy/util/HashUtil.java:30]: it is better to use CRYPTO_ALGORITHM constant to build exception message

File: src/main/resources/hibernate.cfg.xml [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+        <property name="connection.password">admin</property>
+        <property name="show_sql">true</property>
+        <property name="hbm2ddl.auto">create-drop</property>
+
+        <mapping class="mate.academy.model.Movie"/>
+        <mapping class="mate.academy.model.MovieSession"/>
+        <mapping class="mate.academy.model.CinemaHall"/>
+        <mapping class="mate.academy.model.User"/>
+    </session-factory>
+</hibernate-configuration>
```
Comment (Sofasmile) [inline src/main/resources/hibernate.cfg.xml:16]: add empty line at the end of the file

**Review by Sofasmile — CHANGES_REQUESTED** [review]

**Review by Sofasmile — APPROVED** [review]

**Pull request closed without merging**